CMakeLists.txt

```diff
@@ -157,94 +184,204 @@ if(R_LIB)
   list(APPEND SOURCES ${R_SOURCES})
 endif()
 
+# Utility objxgboost used in
+# * xgboost   : R configuration or standard
+# * xgboost4j : JVM_BINDINGS == ON
 add_library(objxgboost OBJECT ${SOURCES})
-
+target_include_directories(gpuxgboost PUBLIC "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>")  
+foreach(dep ${LINK_LIBRARIES})
+  target_include_directories(objxgboost PUBLIC $<TARGET_PROPERTY:${dep},INTERFACE_INCLUDE_DIRECTORIES>)
+  target_compile_definitions(objxgboost PUBLIC $<TARGET_PROPERTY:${dep},INTERFACE_COMPILE_DEFINITIONS>)
+endforeach()
+
+# Note: Some CMake generators won't support libraries without at least 1 source file:
+set(xgboost_null_tmp ${CMAKE_CURRENT_BINARY_DIR}/tmp/xgboost_null.cc)
+set(xgboost_null ${CMAKE_CURRENT_BINARY_DIR}/xgboost_null.cc)
+file(WRITE ${xgboost_null_tmp} "")
+add_custom_command(
+  OUTPUT "${xgboost_null}" # >= 1 non-object file is required on some platforms
+  DEPENDS "${xgboost_null_tmp}"
+  COMMAND ${CMAKE_COMMAND} -E copy_if_different "${xgboost_null_tmp}" "${xgboost_null}"
+  )
 
 # building shared library for R package
 if(R_LIB)
   find_package(LibR REQUIRED)
-
   list(APPEND LINK_LIBRARIES "${LIBR_CORE_LIBRARY}")
   MESSAGE(STATUS "LIBR_CORE_LIBRARY " ${LIBR_CORE_LIBRARY})
 
-  include_directories(
-    "${LIBR_INCLUDE_DIRS}"
-    "${PROJECT_SOURCE_DIR}"
-  )
-
-  # Shared library target for the R package
-  add_library(xgboost SHARED $<TARGET_OBJECTS:objxgboost>)
+  get_cmake_property(_variableNames VARIABLES)
+  foreach (_variableName ${_variableNames})
+    message(STATUS "${_variableName}=${${_variableName}}")
+  endforeach()
+  
+  # Shared library target for the R package:
+  add_library(xgboost SHARED ${xgboost_null} $<TARGET_OBJECTS:objxgboost>)
   target_link_libraries(xgboost ${LINK_LIBRARIES})
+  foreach(target objxgboost xgboost)
+    target_include_directories(${target} PUBLIC "$<BUILD_INTERFACE:${LIBR_INCLUDE_DIRS}>")
+  endforeach()  
+  
   # R uses no lib prefix in shared library names of its packages
   set_target_properties(xgboost PROPERTIES PREFIX "")
 
   setup_rpackage_install_target(xgboost ${CMAKE_CURRENT_BINARY_DIR})
+  
   # use a dummy location for any other remaining installs
   set(CMAKE_INSTALL_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/dummy_inst")
 
-# main targets: shared library & exe
 else()
-  # Executable
-  add_executable(runxgboost $<TARGET_OBJECTS:objxgboost> src/cli_main.cc)
-  set_target_properties(runxgboost PROPERTIES
-    OUTPUT_NAME xgboost
-  )
-  set_output_directory(runxgboost ${PROJECT_SOURCE_DIR})
-  target_link_libraries(runxgboost ${LINK_LIBRARIES})
 
-  # Shared library
-  add_library(xgboost SHARED $<TARGET_OBJECTS:objxgboost>)
+  option(XGBOOST_BUILD_SHARED "Build shared library" ON)
```

Comment:
It's better to use standard `BUILD_SHARED_LIBS`.

Reply:
I thought about that, but in this case it seems we want an option to specifically target the final `xgboost` target, and not the utility `gpuxgboost` target, for example.